### DISCUSSION TITLE
Ambiguous mapping methods processing policy

### DISCUSSION THREAD

**Discussion by mherarsh** [mapstruct/mapstruct#3324] [Q&A]
Hello, everyone!
I have a very obvious task, I want to make a general mapper for certain types, for example I need to make a trim() for all String type fields.

When the Mapper does not use other Mapper there are no problems and everything works fine, but when other Mapper are used it is not clear how to solve the problem.

For example:

There are corresponding classes in Dtos:

```java
class User {
    private String name;
    private UserRole userRole;
    private UserMail userMail;   
   // getters and setters 
}

class UserMail {
    private String email;
   // getters and setters 
}

class UserRole {
    private String roleName;
   // getters and setters 
}

// and dtos --->>> UserDto, UserRoleDto, UserMailDto
```

As I wrote earlier, it is necessary to create a common mapper and use it

```java
@Mapper
interface MapperDefaults {
    default String trim(String value) {
        return value == null ? null : value.trim();
    }
}

@Mapper
interface UserRoleMapper extends MapperDefaults {
    UserRoleDto toDto(UserRole userRole);
}

@Mapper
interface UserMailMapper extends MapperDefaults {
    UserMailDto toDto(UserMail userMail);
}

@Mapper(uses = {UserRoleMapper.class, UserMailMapper.class})
interface UserMapper extends MapperDefaults {
    UserDto toDto(User user);
}

// OR

@Mapper
interface MapperDefaults {
    default String trim(String value) {
        return value == null ? null : value.trim();
    }
}

@Mapper(uses = {MapperDefaults .class})
interface UserRoleMapper {
    UserRoleDto toDto(UserRole userRole);
}

@Mapper(uses = {MapperDefaults .class})
interface UserMailMapper extends MapperDefaults {
    UserMailDto toDto(UserMail userMail);
}

@Mapper(uses = {UserRoleMapper.class, UserMailMapper.class, MapperDefaults .class})
interface UserMapper {
    UserDto toDto(User user);
}
```

The problem is that in this case MapStruct cannot determine which of the trim() methods to use. Obviously these methods are the same. In this case there should be no error, but MapStruct gives the following error: Ambiguous mapping methods found for mapping property "String id" to String: String trim(String value).

The problem can be solved by using Qualifier, but in this case you have to annotate all the methods which makes using common mappers useless.

If there are 2 or more implementations of String trim(String value) it is clear that you can specify to choose via Qualifier otherwise we should have a default method, for example as the @Primary annotation works in Spring Framework.

Please help me with a solution to this question.
Is there such a functionality?
If not, what do you think? Maybe it should be added?
There are so many generic mappers, like for converting dates, strings, enum, etc.

Thank you!

**Comment by filiphr** [comment]
You don't need to extend from util mappers. What you put in `Mapper#uses` is actually not just mappers, but it can be any other class that has static methods in it or other non-mappers that would be instantiated. So in your case I would do the following:

```java
interface MappingUtils {
    static String trim(String value) {
        return value == null ? null : value.trim();
    }
}

@Mapper(uses = MappingUtils.class)
interface UserRoleMapper {
    UserRoleDto toDto(UserRole userRole);
}

@Mapper(uses = MappingUtils.class)
interface UserMailMapper {
    UserMailDto toDto(UserMail userMail);
}

@Mapper(uses = {UserRoleMapper.class, UserMailMapper.class, MappingUtils.class})
interface UserMapper {
    UserDto toDto(User user);
}
```
